flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetAggregateRule.scala

```diff
@@ -33,20 +33,32 @@ class DataSetAggregateRule
       "DataSetAggregateRule")
   {
 
-    def convert(rel: RelNode): RelNode = {
-      val agg: LogicalAggregate = rel.asInstanceOf[LogicalAggregate]
-      val traitSet: RelTraitSet = rel.getTraitSet.replace(DataSetConvention.INSTANCE)
-      val convInput: RelNode = RelOptRule.convert(agg.getInput, DataSetConvention.INSTANCE)
-
-      new DataSetAggregate(
-        rel.getCluster,
-        traitSet,
-        convInput,
-        agg.getNamedAggCalls,
-        rel.getRowType,
-        agg.getInput.getRowType,
-        agg.getGroupSet.toArray)
-      }
+  override def matches(call: RelOptRuleCall): Boolean = {
```

Comment:
That's a good idea. I won't work for all unsupported operators (inner equi-joins are initially Cartesian products + filters before translation rules merge join and filter), but some operations such as distinct aggregates are not rewritten be rules such that we can throw an exception when they are observed.